Mini projects list update

NlightNFotis opened the pull request:
This fixes #4111. Port of a list of starter projects from the old CPROVER wiki
into the in-repo file tracking them.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add more projects to MINI-PROJECTS.md

Port the list of projects available at
https://web.archive.org/web/20150530010030/http://www.cprover.org/wiki/doku.php?id=starter_projects
to the new markdown file inside the repo.

NlightNFotis commented:
Note to reviewers: While porting this list, I found a reference to a lot of names, or WIP notes that I couldn't judge the validity of. These are marked as `(?)` in the file, and I would appreciate if someone could tell me if they are valid or not any more.

Also while porting the list, I removed references to a couple of projects like the DALVIK one, judging that it's no longer relevant. If someone thinks otherwise, I'm happy to add it back.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: df9625add6bc9f0f98705724268b3557e747312f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100110943

NlightNFotis commented:
@tautschnig Could I please get a review on this and get it merged if it looks good?

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+multiple SAT solvers but no command line options to pick a SAT solver. It would be
+nice if there were.
+
+César is working on this one. (?)
```
I think we also have a stalled pull request (#1501 ?) on that one.

peterschrammel replied:
No, #1501 is about Minisat options being passed through, #3243 seems similar.
Cesar is definitely not working on that - so, remove that line.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+César is working on this one. (?)
+
+### Combine Uninitialize Variables Analysis with Alias Analysis
```
Uninitialized

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+`goto-instrument` features an instrumentation to track cases of use-before-define, i.e.
+reading of uninitialised variables. At present, the analysis does not take into account
+aliasing information. As such, the following code will yield a spurious counterexample:
```
I actually think that instrumentation could do with some initial work just to make sure it works at all :-)

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+a program transformation that performs constant folding. ~To ensure soundness in concurrent
+settings, it should be combined with the existing is_threaded analysis.~ (A concurrency-aware
+version of reaching definitions now exists and is awaiting a merge.) Along the way, assertions
+that become trivially true should be eliminated.
```
I believe constant folding is actually implemented to the degree described here.

martin-cs replied:
`goto-analyzer --simplify out.gb --vsd ...` should do everything except the concurrency aspect.  Good concurrent abstract interpretation is non-trivial.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+E.g look at [this paper](https://www.umsec.umn.edu/publications/Partial-Translation-Verification-Untrusted-Code)
+(And please get in touch with Michael Tautschnig, who has done very similar work in FShell).
+
+Rajdeep is now looking into this one. (?)
```
No idea whether that's still a live project.

peterschrammel replied:
I don't think so. I suggest to remove this comment.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+CBMC allows to choose among a selection of SMT solvers to use as solving back end. As these
+tools evolve, CBMC's code for running them may need a fresh look from time to time. After
+a first round of manual experiments, these tests should become part of the routine build
+process on `dkr-build`. (Michael Tautschnig knows how to do this.)
```
Today we've got proper CI, no need to resort do `dkr-build`.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+to execute this trace. This will allow a faster verify/debug cycle. (To a certain
+extent this has been accomplished in FShell. Speak to Michael.)
+
+Pascal is now looking at this one. (?)
```
I guess there is an actual product for Java nowadays :-)

peterschrammel replied:
Yes... can be removed.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+support for dynamic analysis, calls to external functions and possibly even mixed concrete
+and symbolic execution. It would also be good to be able to interpret traces found by BMC
+to catch mismatches / bugs between the logical model and the bit blasting models. This
+would also be useful for more advanced CEGAR like techniques (also see below).
```
I have no idea what the actual state of the interpreter is today.

peterschrammel replied:
I'd propose a project to implement an interpreter using `goto-symex`.

martin-cs replied:
Using the abstract interpreter would also be possible.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+* Use path exploration as the witness generation part of an ACDL process.
+* Perform precondition inference, possibly using ACDL.
+
+See the work Björn and Alex have done with path-symex. (?)
```
We do now have a single-path symex mode, but the ideas above might still be relevant.

peterschrammel replied:
Unclear what the points above mean, but there is certainly work to be done around the single-path symex, including robustness, efficiency and search strategies.

martin-cs replied:
I believe the original suggestion (I have a feeling I wrote a number of these) was using ACDL-like reasoning to "learn" invariants from failed symex runs (using UNSAT cores and the like) and add these in to the process, giving something like IMPACT but without the covering relations.

tautschnig commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+## Interact with GDB
+
+GDB implements a client-server protocol with a GDB instance acting as a server and the
```
There is an open PR on that one (#3983).

peterschrammel replied:
I don't think that covers all the points below.

tautschnig reviewed (commented):
I think this should get a thorough review from @chrisr-diffblue @peterschrammel and whoever else has a good overview of what the status of various bits of work is.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+### Error Explanation
+
+Re-implement the algorithm described in [this paper](http://www.kroening.com/papers/STTT-error-explanation-2005.pdf)
```
This is partly implemented in the fault localizer.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+### Interpreter as a GDB server
+
+An enabling tech, this would allow us to run the interpretter as a debugger back end. Combine
```
interpreter

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+GDB implements a client-server protocol with a GDB instance acting as a server and the
+GUI as a client. (The documentation for that protocol may be found [here](http://www.sourceware.org/gdb/current/onlinedocs/gdb/GDB_002fMI.html).)
+By implementing this protocol in various place within CProver, there are various
```
places

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+debugger that can talk to GDB to be used to graphically explore the trace. Particularly
+valuable for concurrent programs in which just reading the trace could be very hard.
+
+Pascal has something like this implemented for Eclipse. (?)
```
An implementation of that is in the trace debugger here: https://github.com/diffblue/eclipse-cbmc

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+for catching encoding bugs in CProver and ideally should be run after the decode from the solver
+all of the time.
+
+Michael has a partial implementation of this. (?)
```
Really, @tautschnig ?

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+Implement [k-liveness](http://www.cs.utexas.edu/~hunt/FMCAD/FMCAD12/013.pdf)
+for software-liveness checking.
+
+Hongyi is now looking at this (?)
```
Not anymore - remove this.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+so possibly some combination of this, trace generation, DRD / Helgrind and then
+refinement of which traces to pick? Perhaps this is even something ACDL like?
+
+### CORAL / Q
```
I'd propose to implement something like stratified inlining, for example.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+### Known Bugs on the Web
+
+Turn the list of `KNOWNBUG`S from the CBMC regressions into a webpage of known bugs
```
Webpage means github issue tracker

peterschrammel replied:
We should also use the information on fixed bugs that we get from CI.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+the [GCC C Torture tests](https://gcc.gnu.org/onlinedocs/gccint/C-Tests.html) with `goto-cc` and a
+`goto-program` interpreter (mild dependency on the interpreter-related task above).
+
+### Unspecified C Behaviours
```
That's a subset of the Coding Standards Checker.

peterschrammel commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+
+It would probably be worth asking Nando de Freitas' advice on the machine learning side of things.
+
+Ruben is now looking at this one. (?)
```
Not anymore - remove this line.

peterschrammel commented on MINI-PROJECTS.md:
Cesar and Kurt have done work on that.

peterschrammel commented on MINI-PROJECTS.md:
An old prototype of something similar is here: https://github.com/peterschrammel/cbmc/commits/progres-exceptions

peterschrammel commented on MINI-PROJECTS.md:
path-symex

peterschrammel commented on MINI-PROJECTS.md:
@tautschnig, yes?

TGWDB commented:
@NlightNFotis Is this still going to be worked on and maybe merged?

martin-cs commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+address is a multiple of the word length (i.e. 4 bytes or 8 bytes). This project
+would add a new instrumentation option for `cbmc` and `goto-instrument` that checks
+each pointer before dereference and makes sure it's correctly aligned.
+
```
This is a good way to get into `goto_check` and how instrumentation of asserts work.

martin-cs commented on MINI-PROJECTS.md:
````diff
@@ -104,6 +104,314 @@ share a bound.
+  return 0;
+}
+```
+
````
This should be pretty straight-forward to do with `goto-analyser --vsd` and might even be a good intro to that code.

martin-cs commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+### CryptoMinisat
+
+Try [CryptoMinisat](https://github.com/msoos/cryptominisat) vs Minisat.
+
```
I believe this is now implemented.

martin-cs commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+others may be best to be instrumentation (for example handling the unspecified nature
+of the order of evaluation of function arguments).
+
+Talk to Michael Tautschnig or Martin Brain for this item.
```
I am, of course, happy to chat to anyone about any of these but I don't recall anything specific about this one so maybe I shouldn't be mentioned directly?

martin-cs commented on MINI-PROJECTS.md:
```diff
@@ -104,6 +104,314 @@ share a bound.
+work, etc. Rather than trying to manually pick the best strategy from these, it would be good
+to machine learn them.
+
+It would probably be worth asking Nando de Freitas' advice on the machine learning side of things.
```
Although I am sure that he has many insightful and helpful things to say on the subject, the specific link between him and the bulk of the CPROVER developers (being at the same institution, one floor different, etc.) has gone so maybe we shouldn't explicitly name check him here?

martin-cs commented on MINI-PROJECTS.md:
This is `goto-analyzer`.  We can remove this project suggestion -- turns out to not be so mini after all!

martin-cs commented on MINI-PROJECTS.md:
@tautschnig has implemented the CSmith part.

martin-cs reviewed (approved):
It would be great to see this patch merged but it does need a few updates.

NlightNFotis commented:
@TGWDB Yes, I intend to chase it up and clean this up soon (as soon as I get some time).

@martin-cs Thanks for the suggestions and the comments, will incorporate them when I find some time 👍🏻

NlightNFotis committed:
Address review comments, and appease the Markdown linter.

NlightNFotis commented:
Hi @peterschrammel and @tautschnig, could you take a look at this at your latest convenience?

I've implemented most of the comments, and changed the file a bit to conform to markdown listing standards.

Other than this, it looks good to go to me, what do you think?
